bundles/org.openhab.automation.jrubyscripting/src/main/java/org/openhab/automation/jrubyscripting/internal/watch/JRubyDependencyTracker.java

```diff
@@ -0,0 +1,106 @@
+/**
+ * Copyright (c) 2010-2022 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.automation.jrubyscripting.internal.watch;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Consumer;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.openhab.automation.jrubyscripting.internal.JRubyScriptEngineFactory;
+import org.openhab.core.automation.module.script.ScriptDependencyTracker;
+import org.openhab.core.service.AbstractWatchService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Tracks Ruby dependencies
+ *
+ * @author Cody Cutrer - Initial contribution
+ */
+@NonNullByDefault
+public class JRubyDependencyTracker implements ScriptDependencyTracker {
```

Comment:
This could be a service (component) itself which then is injected into the script engine factory (like in js). I didn't check where the individual configurations values come from, but likely they can be derived from the ConfigAdmin.

Reply:
I tried _really_ hard to get this to be an OSGi component. I actually started by just copying the JS files. But I couldn't get around the circular dependence - `JRubyScriptEngineFactory` needs to return the dependency tracker. But the dependency tracker needs access to the config. On the one hand, like you say, I could just use `ConfigAdmin` directly to get the current configuration values, but the raw configuration is wrapped by Java both to provide non-constant defaults (well, it is theoretically constant, but dependent on values from the JRuby bundle), as well as providing some post-processing to the values. I definitely didn't want to attempt to duplicate that.

I also considered just having the `JRubyScriptEngineFactory` implement `ScriptDependencyTracker` like you said, but I wanted to keep it separate in hopes of eliminating the bulk of this class by inheriting from `AbstractScriptDependencyTracker` as discussed in the other comment.

Another option I've thought of, but didn't really explore, is having the configuration itself be an independent component that both `JRubyScriptEngineFactory` and `JRubyDependencyTracker` depend on. If you're adamant that you really want this to be an OSGi component, I can explore that.

Reply:
I would suggest to keep it like it is now and refactor it later when the core watch service is changed. As a result it should also be possible to easily adapt the `AbstractScriptDependencyTracker` to watch for more than one directory. Since that one needs refactoring anyway with the new watch service, I would prefer not to touch it before the release because time is a bit short.